specs/agents/tracing-instrumentation-messaging.md

```diff
@@ -0,0 +1,103 @@
+  - `transaction.type`: `messaging`
+- Spans: 
+  - `span.type`: `messaging` 
+  - `span.subtype`: the name of the framework - e.g. `jms`, `kafka`, `rabbitmq` 
+  - `span.action`: `send`, `receive` or `poll`
+  
+### Naming
+
+Transaction and span names may* follow this pattern: `<MSG-FRAMEWORK> SEND/RECEIVE/POLL to/from <QUEUE-NAME>`.
+Examples:
+- `JMS SEND to MyQueue`
+- `RabbitMQ RECEIVE from MyQueue`**
+- `RabbitMQ POLL from MyExchange`**
+
+Agents may deviate from this pattern, as long as they ensure a proper cardinality is maintained, that is- neither too low nor too high. 
+For example, agents may choose to name all transactions/spans reading-from/sending-to temporary queues equally. 
+On the other end, agents may choose to append a cardinality-increasing factor to the name, like the routing key in RabbitMQ.
+
+\* Java agent's instrumentation for Kafka does not follow this pattern at the moment.
+
+#### \** RabbitMQ naming specifics 
+
+In RabbitMQ, queues are only relevant in the receiver side, so the exchange name is used instead for sender spans.
+When the default exchange is used (denoted with an empty string), it should be replaced with `<default>`. 
+
+Agents may add an opt-in config to append the routing key to the name as well, for example: `RabbitMQ RECEIVE from MyExchange\58D7EA987`.
+
+For RabbitMQ transaction and polling spans, the queue name is used instead, whenever available (i.e. when the polling yields a message).
+
+### Context fields
+
+- **`context.message.queue.name`**: optional for `messaging` spans and transactions. Indexed as keyword. Wherever the broker terminology 
+uses "topic", this field will contain the topic name. In RabbitMQ, whenever the queue name is not available, use exchange name instead.
+- **`context.message.age.ms`**: optional for message/record receivers only (transactions or spans). 
+A numeric field indicating the message's age in milliseconds. Relevant for transactions and 
+`receive` spans that receive valid messages. There is no accurate definition as to how this is calculated. If the messaging framework 
+provides a timestamp for the message- agents may use it. Otherwise, the sending agent can add a timestamp _indicated as milliseconds since 
+epoch UTC_ to the message's metadata to be retrieved by the receiving agent. If a timestamp is not available- agents should omit this field. 
+Clock skews between agents are ignored, unless the calculated age (receive-timestamp minus send-timestamp) is negative, in which case the 
+agent should report 0 for this field.
+- **`context.message.routing-key`**: optional. Use only where relevant. Currently only RabbitMQ.
+
+#### Transaction context fields
+
+- **`context.message.body`**: similar to HTTP requests' `context.request.body`- only fill in messaging-related **transactions** (ie 
+incoming messages creating a transaction) and not for outgoing messaging spans. 
+   - Capture only when `ELASTIC_APM_CAPTURE_BODY` config is set to `true`.
```

Comment:
We should update the documentation in central config to make it clear that this not only affects HTTP but also messaging?

Do we want or need to support use cases where users want to capture HTTP request bodies but no message bodies or vice-versa?

How does `capture_body_content_types` play into this?

Reply:
> We should update the documentation in central config to make it clear that this not only affects HTTP but also messaging?

We should and we already do. The central config description reads: `For transactions that are HTTP requests, the agent can optionally capture the request body (e.g. POST variables). For transactions that are initiated by receiving a message from a message broker, the agent can capture the textual message body.`

> Do we want or need to support use cases where users want to capture HTTP request bodies but no message bodies or vice-versa?

For now I think we are good with one setting. I assume a service will typically create transactions either mostly based on HTTP requests or mostly based on messages, but not an equal mix of both. In addition, the consideration is the same with regard to the tradeoff of more data for storage (and other) overhead.
If we see a request for separate configs in the future, we can reiterate.

> How does capture_body_content_types play into this?

There is no properly defined close set of content types in messaging systems. In Kafka we use the default textual representation of the key and value objects. In JMS we only capture body for text messages based on the JMS `Message` object type. So I suggest we do not attempt to use this config for messaging.